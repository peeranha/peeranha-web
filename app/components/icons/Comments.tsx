import React from 'react';
import IconComponent, { IconProps } from './IconComponent';

const Comments: React.FC<IconProps> = (props): JSX.Element => (
  <IconComponent
    {...props}
    id="comments"
    fill="curentColor"
    viewBox="0 0 24 24"
    size={[24, 24]}
  >
    <g transform="translate(0.5, 0.5)">
      <path
        fill="none"
        stroke="#000000"
        strokeWidth="1"
        d="M7.799,16.601
	      C9.092,18.61,11.826,20,15,20c0.59,0,1.163-0.051,1.716-0.142L21,22v-4.04c1.241-1.057,2-2.44,2-3.96
	      c0-1.552-0.792-2.961-2.081-4.027"
        strokeLinejoin="miter"
        strokeLinecap="butt"
      />
      <path
        fill="none"
        stroke="#000000"
        strokeWidth="1"
        strokeLinecap="square"
        d="M11,1C5.477,1,1,4.582,1,9
	      c0,1.797,0.75,3.45,2,4.785V19l4.833-2.416C8.829,16.85,9.892,17,11,17c5.523,0,10-3.582,10-8S16.523,1,11,1z"
        strokeLinejoin="miter"
      />
    </g>
  </IconComponent>
);

export default Comments;
