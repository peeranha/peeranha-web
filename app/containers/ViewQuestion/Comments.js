import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { FormattedMessage } from 'react-intl';

import {
  TEXT_PRIMARY,
  BORDER_SECONDARY,
  BORDER_PRIMARY,
  BORDER_TRANSPARENT,
} from 'style-constants';

import editSmallIcon from 'images/editSmallIcon.svg?inline';
import deleteSmallIcon from 'images/deleteSmallIcon.svg?inline';
import blockSmallIcon from 'images/blockSmallIcon.svg?inline';

import Span from 'components/Span';
import Textarea from 'components/Textarea';

import Button, { BlockButton } from './Button';
import UserInfo from './UserInfo';
import CommentOptions from './CommentOptions';
import CommentForm from './CommentForm';

import messages from './messages';

import {
  SAVE_COMMENT_FORM,
  SAVE_COMMENT_BUTTON,
  COMMENT_TYPE,
} from './constants';

const CommentManage = styled.div`
  opacity: 0;

  button {
    padding-left: 15px !important;

    ${Span} {
      font-size: 14px;
    }
  }
`;

const CommentViewStyled = styled.li`
  position: relative;
`;

const CommentEditStyled = styled.li`
  ${Textarea} {
    height: 90px;
  }
`;

const CommentsStyled = styled.ul`
  ${CommentViewStyled} {
    border: 1px solid ${BORDER_SECONDARY};
    padding: 5px 15px 8px 15px;

    :first-child {
      border-top-left-radius: 3px;
      border-top-right-radius: 3px;
    }

    :last-child {
      border-bottom-left-radius: 3px;
      border-bottom-right-radius: 3px;
    }

    :not(:last-child) {
      border-bottom-color: ${BORDER_TRANSPARENT};
    }

    :hover {
      border-color: ${BORDER_PRIMARY};

      ${CommentManage} {
        opacity: 1;
      }
    }
  }
`;

const CommentEdit = /* istanbul ignore next */ ({
  answerId,
  id,
  content,
  translations,
  saveCommentLoading,
  saveComment,
  toggleView,
}) => (
  <CommentEditStyled className="my-4">
    <CommentForm
      form={`${SAVE_COMMENT_FORM}_${answerId}${id}`}
      comment={content}
      submitButtonId={SAVE_COMMENT_BUTTON}
      submitButtonName={translations[messages.saveButton.id]}
      sendCommentLoading={saveCommentLoading}
      sendComment={saveComment}
      answerId={answerId}
      commentId={id}
      toggleView={toggleView}
    />
  </CommentEditStyled>
);

const CommentView = /* istanbul ignore next */ item => (
  <CommentViewStyled>
    <div className="d-flex justify-content-between align-items-center">
      <UserInfo
        type={COMMENT_TYPE}
        avatar={item.userInfo.ipfs_avatar}
        name={item.userInfo.display_name}
        rating={item.userInfo.rating}
        account={item.userInfo.user}
        postTime={item.post_time}
        locale={item.locale}
      />

      <CommentManage>
        <Button
          show={item.isItWrittenByMe}
          params={{
            ...item.buttonParams,
            commentId: item.id,
            whowasvoted: item.userInfo.user,
          }}
          onClick={() => item.toggleView(!item.isView)}
        >
          <Span className="d-flex align-items-center" color={TEXT_PRIMARY}>
            <img className="mr-1" src={editSmallIcon} alt="icon" />
            <FormattedMessage {...messages.editButton} />
          </Span>
        </Button>

        <Button
          show={item.isItWrittenByMe}
          id={`comment__${item.answerId}${item.id}`}
          params={{
            ...item.buttonParams,
            commentId: item.id,
            whowasvoted: item.userInfo.user,
          }}
          onClick={item.deleteComment}
        >
          <Span className="d-flex align-items-center" color={TEXT_PRIMARY}>
            <img className="mr-1" src={deleteSmallIcon} alt="icon" />
            <FormattedMessage {...messages.deleteButton} />
          </Span>
        </Button>

        <BlockButton
          show
          id={`comment_vote_to_delete_${item.answerId}${item.id}`}
          isVotedToDelete={item.isVotedToDelete}
          isItWrittenByMe={item.isItWrittenByMe}
          onClick={item.voteToDelete}
          params={{
            ...item.buttonParams,
            commentId: item.id,
            whowasvoted: item.userInfo.user,
          }}
        >
          <Span className="d-flex align-items-center" color={TEXT_PRIMARY}>
            <img className="mr-1" src={blockSmallIcon} alt="icon" />
            <FormattedMessage {...messages.voteToDelete} />
          </Span>
        </BlockButton>
      </CommentManage>
    </div>

    <div dangerouslySetInnerHTML={{ __html: item.content }} />
  </CommentViewStyled>
);

const Comment = /* istanbul ignore next */ item => {
  const [isView, toggleView] = useState(true);

  return (
    <React.Fragment>
      {!isView ? (
        <CommentEdit toggleView={toggleView} {...item} />
      ) : (
        <CommentView isView={isView} toggleView={toggleView} {...item} />
      )}
    </React.Fragment>
  );
};

const Comments = /* istanbul ignore next */ props => {
  const DEFAULT_COMMENTS_NUMBER = 3;
  const [isAllCommentsView, changeCommentsView] = useState(false);

  const commentsNum = !isAllCommentsView ? DEFAULT_COMMENTS_NUMBER : undefined;

  return (
    <div>
      <CommentsStyled className="my-3">
        {props.comments
          .slice(0, commentsNum)
          .map(item => (
            <Comment {...item} {...props} key={`${COMMENT_TYPE}${item.id}`} />
          ))}
      </CommentsStyled>

      <CommentOptions
        form={props.form}
        submitButtonId={props.submitButtonId}
        submitButtonName={props.submitButtonName}
        sendCommentLoading={props.sendCommentLoading}
        sendComment={props.sendComment}
        answerId={props.answerId}
        changeCommentsView={changeCommentsView}
        isAllCommentsView={isAllCommentsView}
        commentsNumber={props.comments.length - DEFAULT_COMMENTS_NUMBER}
      />
    </div>
  );
};

Comments.propTypes = {
  comments: PropTypes.array,
  form: PropTypes.string,
  submitButtonId: PropTypes.string,
  submitButtonName: PropTypes.string,
  sendCommentLoading: PropTypes.bool,
  sendComment: PropTypes.func,
  answerId: PropTypes.string,
};

CommentEdit.propTypes = {
  answerId: PropTypes.string,
  id: PropTypes.string,
  content: PropTypes.string,
  translations: PropTypes.object,
  saveCommentLoading: PropTypes.bool,
  saveComment: PropTypes.func,
  toggleView: PropTypes.func,
};

export { Comment, CommentEdit, CommentView, Comments };
export default React.memo(Comments);
