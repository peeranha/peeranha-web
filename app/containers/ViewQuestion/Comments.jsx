import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { FormattedMessage } from 'react-intl';

import {
  BORDER_SECONDARY,
  BORDER_PRIMARY,
  BORDER_TRANSPARENT,
} from 'style-constants';

import editSmallIcon from 'images/editSmallIcon.svg?inline';
import deleteSmallIcon from 'images/deleteSmallIcon.svg?inline';
import blockSmallIcon from 'images/blockSmallIcon.svg?external';

import { getUserAvatar } from 'utils/profileManagement';

import Span from 'components/Span';
import Icon from 'components/Icon';
import Textarea from 'components/Textarea';

import Button from './Button';
import UserInfo from './UserInfo';
import CommentOptions from './CommentOptions';
import CommentForm from './CommentForm';

import messages from './messages';

import {
  SAVE_COMMENT_FORM,
  SAVE_COMMENT_BUTTON,
  COMMENT_TYPE,
} from './constants';

import AreYouSure from './AreYouSure';

const CommentManage = styled.div`
  display: flex;
  opacity: 0;

  button {
    margin-left: 13px;
    font-size: 14px;

    > *:last-child {
      margin-left: 4px;
    }
  }

  @media only screen and (max-width: 576px) {
    position: absolute;
    top: 0;
    right: 0;
  }
`;

const CommentEditStyled = styled.li`
  ${Textarea} {
    height: 90px;
  }
`;

const CommentsStyled = styled.ul`
  margin-top: 30px;

  ${CommentEditStyled} {
    padding: 0;
    border: none;
  }

  li {
    border: 1px solid ${BORDER_SECONDARY};
    padding: 9px 28px 9px 38px;

    :first-child {
      border-top-left-radius: 3px;
      border-top-right-radius: 3px;
    }

    :last-child {
      border-bottom-left-radius: 3px;
      border-bottom-right-radius: 3px;
    }

    :not(:last-child) {
      border-bottom-color: ${BORDER_TRANSPARENT};
    }

    :hover {
      border-color: ${BORDER_PRIMARY};

      ${CommentManage} {
        opacity: 1;
      }
    }
  }

  @media only screen and (max-width: 576px) {
    li {
      padding: 5px 15px;
    }
  }
`;

const CommentEdit = ({
  answerId,
  id,
  content,
  translations,
  saveCommentLoading,
  saveComment,
  toggleView,
}) => (
  <CommentEditStyled className="my-4">
    <CommentForm
      form={`${SAVE_COMMENT_FORM}_${answerId}${id}`}
      comment={content}
      submitButtonId={SAVE_COMMENT_BUTTON}
      submitButtonName={translations[messages.saveButton.id]}
      sendCommentLoading={saveCommentLoading}
      sendComment={saveComment}
      answerId={answerId}
      commentId={id}
      toggleView={toggleView}
    />
  </CommentEditStyled>
);

/* eslint react/no-danger: 0 */
const CommentView = item => (
  <li>
    <div className="d-flex justify-content-between align-items-center position-relative">
      <UserInfo
        type={COMMENT_TYPE}
        avatar={getUserAvatar(item.userInfo.ipfs_avatar)}
        name={item.userInfo.display_name}
        rating={item.userInfo.rating}
        account={item.userInfo.user}
        postTime={item.post_time}
        locale={item.locale}
      />

      <CommentManage>
        <Button
          show={item.isItWrittenByMe}
          params={{
            ...item.buttonParams,
            commentId: item.id,
            whowasvoted: item.userInfo.user,
          }}
          onClick={() => item.toggleView(!item.isView)}
        >
          <img src={editSmallIcon} alt="icon" />
          <FormattedMessage {...messages.editButton} />
        </Button>

        <div id={`delete-comment-${item.answerId}${item.id}`}>
          <AreYouSure
            submitAction={item.deleteComment}
            Button={({ onClick }) => (
              <Button
                show={item.isItWrittenByMe}
                id={`delete-comment-${item.answerId}${item.id}`}
                params={{
                  ...item.buttonParams,
                  commentId: item.id,
                  whowasvoted: item.userInfo.user,
                }}
                onClick={onClick}
                disabled={item.ids.includes(
                  `delete-comment-${item.answerId}${item.id}`,
                )}
              >
                <img src={deleteSmallIcon} alt="icon" />
                <FormattedMessage {...messages.deleteButton} />
              </Button>
            )}
          />
        </div>

        <Button
          show={!item.isItWrittenByMe}
          id={`comment_vote_to_delete_${item.answerId}${item.id}`}
          params={{
            ...item.buttonParams,
            commentId: item.id,
            whowasvoted: item.userInfo.user,
          }}
          onClick={item.voteToDelete}
          disabled={item.ids.includes(
            `comment_vote_to_delete_${item.answerId}${item.id}`,
          )}
          isVotedToDelete={item.votingStatus.isVotedToDelete}
        >
          <Icon icon={blockSmallIcon} width="12" />
          <FormattedMessage {...messages.voteToDelete} />
        </Button>
      </CommentManage>
    </div>

    <Span
      fontSize="16"
      lineHeight="20"
      className="d-block mb-2"
      dangerouslySetInnerHTML={{ __html: item.content }}
    />
  </li>
);

const Comment = item => {
  const [isView, toggleView] = useState(true);

  return (
    <React.Fragment>
      {!isView ? (
        <CommentEdit toggleView={toggleView} {...item} />
      ) : (
        <CommentView isView={isView} toggleView={toggleView} {...item} />
      )}
    </React.Fragment>
  );
};

const Comments = props => {
  const isPhone = window.screen.width <= 576;
  const DEFAULT_COMMENTS_NUMBER = isPhone ? 0 : 3;
  const [isAllCommentsView, changeCommentsView] = useState(false);

  const commentsNum = !isAllCommentsView ? DEFAULT_COMMENTS_NUMBER : undefined;

  return (
    <div>
      {props.comments.length > 0 && (
        <CommentsStyled>
          {props.comments
            .slice(0, commentsNum)
            .map(item => (
              <Comment {...item} {...props} key={`${COMMENT_TYPE}${item.id}`} />
            ))}
        </CommentsStyled>
      )}

      <CommentOptions
        form={props.form}
        submitButtonId={props.submitButtonId}
        submitButtonName={props.submitButtonName}
        sendCommentLoading={props.sendCommentLoading}
        sendComment={props.sendComment}
        answerId={props.answerId}
        changeCommentsView={changeCommentsView}
        isAllCommentsView={isAllCommentsView}
        commentsNumber={props.comments.length - DEFAULT_COMMENTS_NUMBER}
      />
    </div>
  );
};

Comments.propTypes = {
  comments: PropTypes.array,
  form: PropTypes.string,
  submitButtonId: PropTypes.string,
  submitButtonName: PropTypes.string,
  sendCommentLoading: PropTypes.bool,
  sendComment: PropTypes.func,
  answerId: PropTypes.string,
};

CommentEdit.propTypes = {
  answerId: PropTypes.string,
  id: PropTypes.string,
  content: PropTypes.string,
  translations: PropTypes.object,
  saveCommentLoading: PropTypes.bool,
  saveComment: PropTypes.func,
  toggleView: PropTypes.func,
};

export { Comment, CommentEdit, CommentView, Comments };
export default React.memo(Comments);
