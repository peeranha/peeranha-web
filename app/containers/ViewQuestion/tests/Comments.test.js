import Comments, { Comment } from '../Comments';

describe('Comments', () => {
  it('test1, Comments', () => {
    const props = {
      comments: [
        {
          id: 1,
        },
        {
          id: 2,
        },
      ],
    };

    expect(Comments(props)).toMatchSnapshot();
  });

  it('test2, Comment', () => {
    const props = {
      userInfo: {},
    };

    expect(Comment(props)).toMatchSnapshot();
  });
});
