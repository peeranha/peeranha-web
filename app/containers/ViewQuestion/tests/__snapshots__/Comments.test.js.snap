// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="comment-content"
>
  <div>
    <ChangesHistory />
    <p
      className="comment-text"
      dangerouslySetInnerHTML={
        Object {
          "__html": undefined,
        }
      }
    />
  </div>
  <p
    className="option-edit"
  >
    <Button
      buttonParams={
        Object {
          "commentId": undefined,
          "whowasvoted": undefined,
        }
      }
    />
    <Button
      buttonId="comment__undefinedundefined"
      buttonParams={
        Object {
          "commentId": undefined,
          "whowasvoted": undefined,
        }
      }
    />
    <Button
      buttonId="comment_vote_to_delete_undefinedundefined"
      buttonParams={
        Object {
          "commentId": undefined,
          "whowasvoted": undefined,
        }
      }
      show={true}
    />
  </p>
</div>
`;

exports[` 2`] = `
<div
  className="save-edited-form"
>
  <Connect(ReduxForm)
    form="app_ViewQuestion_SAVE_COMMENT_FORM_undefinedundefined"
    submitButtonId="app_ViewQuestion_SAVE_COMMENT_BUTTON"
  />
</div>
`;

exports[`Comments CommentVision answerid != item.answerId && commentid != item.id 1`] = `
<CommentView
  answerId={11111}
  editCommentState={
    Object {
      "answerid": 1,
      "commentid": 1,
    }
  }
  id={11111}
/>
`;

exports[`Comments CommentVision answerid == item.answerId && commentid == item.id 1`] = `
<CommentEdit
  answerId={1}
  editCommentState={
    Object {
      "answerid": 1,
      "commentid": 1,
    }
  }
  id={1}
/>
`;

exports[`Comments test1, Comments 1`] = `
<div
  className="comments"
>
  <Comment
    comments={
      Array [
        Object {
          "id": 1,
          "userInfo": Object {},
        },
        Object {
          "id": 2,
          "userInfo": Object {},
        },
      ]
    }
    id={1}
    userInfo={Object {}}
  />
  <Comment
    comments={
      Array [
        Object {
          "id": 1,
          "userInfo": Object {},
        },
        Object {
          "id": 2,
          "userInfo": Object {},
        },
      ]
    }
    id={2}
    userInfo={Object {}}
  />
</div>
`;

exports[`Comments test2, Comment 1`] = `
<div
  className="comment-body"
>
  <UserInfo />
  <CommentVision
    userInfo={Object {}}
  />
</div>
`;
